velox/docs/develop/testing/spark-query-runner.rst

```diff
@@ -0,0 +1,75 @@
+==================
+Spark Query Runner
+==================
+
+Introduction
+------------
+
+The Spark Query Runner is a tool designed to facilitate the testing of Velox.
+It achieves this by executing SQL queries on Apache Spark and comparing the results
+with those produced by Velox. It helps ensure the correctness of Velox's computing
+against Spark and provides a method for identifying potential issues in Velox's
+implementation. Currently Spark-3.5.1 is used in Velox.
+
+How It Works
+------------
+
+The Spark Query Runner operates by executing given SQL queries on both Spark and Velox.
+The results from both systems are then compared to check for any differences.
+If the results match, it indicates that Velox is producing the correct output.
+If the results differ, it suggests a potential issue in Velox that needs to be
+investigated.
+
+Since Spark 3.4, Spark Connect has introduced a decoupled client-server architecture
+for Spark that allows remote connectivity to Spark clusters. From the client
+perspective, Spark Connect mostly behaves as any other gRPC client, which is polyglot
+and and cross-platforms. During execution, the Spark Connect endpoint embedded on the
+Spark Server receives and translates unresolved logical plans into Spark’s logical plan
+operators. From there, the standard Spark execution process kicks in, ensuring that
+Spark Connect leverages all of Spark’s optimizations and enhancements. Results are
+streamed back to the client through gRPC as Apache Arrow-encoded row batches.
+
+In the Spark Query Runner, we use Spark Connect to submit queries to Spark and fetch
+the results back to Velox for comparison. The steps for this process are as follows:
+
+1. Provide the Spark SQL query to be executed. The query could be generated from Velox
+   plan node or manually written.
+2. Create a protobuf message `ExecutePlanRequest` from the SQL query. The protocols
+   used by Spark Connect are defined in `Apache Spark <https://github.com/apache/spark/tree/v3.5.1/connector/connect/common/src/main/protobuf/spark/connect>`_.
+3. Submit the message to SparkConnectService through gRPC API `ExecutePlan`.
+4. Fetch Spark's results from execution response. Results are in Arrow IPC stream format,
+   and can be read as Arrow RecordBatch by `arrow::ipc::RecordBatchReader`.
+5. Convert Arrow RecordBatch as Velox vector for comparison with Velox's results.
+
+Usage
+-----
+
+To use the Spark Query Runner, you will need to deploy an executable Spark and start the
```

Comment:
Do we have step-by-instructions for this somewhere? I'd like to give it try.

Reply:
It could be tried quickly in the docker image `ghcr.io/facebookincubator/velox-dev:spark-server` which contains an executable spark and the needed script. 

Spark can also be deployed manually following the steps in https://github.com/facebookincubator/velox/blob/main/scripts/spark-container.dockerfile through downloading and configuring.